feat(extra-natives/five): add light natives

Ekinoxx0 opened the pull request:
### Goal of this PR
This PR introduces several new natives to create, delete, and manage all properties of scripted lights. This enables more flexible lighting setups and access to features that were not previously available through the traditional `DRAW_LIGHT_...` natives, such as texture projection, capsule shapes, and volumetric flags.

![h9vY2odmrp](https://github.com/user-attachments/assets/a7323944-d3ac-4890-84fb-42babcc95fe9)
![ne9Redg3TE](https://github.com/user-attachments/assets/560c6bb7-9328-4c0c-a49d-d49e37f62145)

### How is this PR achieving the goal

Instead of using a per-frame draw approach, this PR implements a handle-based system to manage lights. With the `CREATE_LIGHT` and `DELETE_LIGHT` natives, users can create lights that persist across frames, improving both performance and usability.

Persisting light objects means they are reused every frame, instead of being recreated, using a draw-based native would have required a large number of arguments to expose all the fields of the `CLight` class.

This PR is a draft because it introduces a significant number of new natives.
I’m looking for feedback on the implementation approach and suggestions for improvement before finalizing the design.

Implementation note:
The everyFrame hook seems weird at first reading, but I didn't find any better way to implement it.
Using fwEvent OnGameFrame doesn't render the lights properly.
We need too add lights just after natives finished executing, but before it's cleaned by the game, this hook was the simpliest way I found.

### This PR applies to the following area(s)
<!-- Add any that applies, e.g.: FiveM, RedM, Server, Natives, FxDK, ScRT: Lua, ScRT: C#, ScRT: JS, etc. -->
FiveM Natives


### Successfully tested on
<!-- Add any that is applicable, remove any that aren't. -->

**Game builds:** 1604 to 3258
**Platforms:** Windows

```lua
CreateThread(function()
	for i = 0, 1024 do DeleteLight(i) end -- This could be replaced with a mission clean-up thing ?

	local color = { r = 255, g = 255, b = 0 }
	local intensity = 1.0
	local radius = 1.0

	local light = CreateLight(1, (1<<12), 0.0, 0.0, 0.0, color.r, color.g, color.b, intensity)
	SetLightRadius(light, radius)

	CreateThread(function()
		while true do
			local camPos = GetFinalRenderedCamCoord()
			local endPoint, normal = GetWorldCoordFromScreenCoord(GetDisabledControlNormal(0, 239), GetDisabledControlNormal(0, 240))
			endPoint = endPoint + normal * 10.0

			local rayHandle = StartExpensiveSynchronousShapeTestLosProbe(camPos.x, camPos.y, camPos.z, endPoint.x, endPoint.y, endPoint.z, -1, PlayerPedId(), 0)
			local _, hit, worldPosition = GetShapeTestResultIncludingMaterial(rayHandle)

			if hit then
				SetLightCoords(light, worldPosition.x, worldPosition.y, worldPosition.z)
			end
			Wait(0)
		end
	end)
end)
```


### Checklist
<!-- Mark all points with x that apply, i.e.: [x]. -->

- [x] Code compiles and has been tested successfully.
- [x] Code explains itself well and/or is documented.
- [x] My commit message explains what the changes do and what they are for.
- [x] No extra compilation warnings are added by these changes.


### Fixes issues
<!-- Add any issue that this PR fixes with: `fixes #123`, `resolves #234`, `closes #345`. -->
Resolve https://forum.cfx.re/t/more-drawblanklight-natives-drawvolumetricspotlight-drawcapsule/5327113

ook3D commented:
> My only thought is about how we can manage effects like bright flashes that only happen for a short time. 

this is a specific light parameter called Flashiness, should be possible to add a similar native for it too

Lucas7yoshi commented:
> > My only thought is about how we can manage effects like bright flashes that only happen for a short time.
> 
> this is a specific light parameter called Flashiness, should be possible to add a similar native for it too

alternatively they may mean just doing a quick blip of light but you could just create, wait, delete?

Ekinoxx0 commented:
**Modified the PR:**

* Removed the handle system (`CLightManager` class)
* Removed the `CREATE_LIGHT` and `REMOVE_LIGHT` natives
* Eliminated the need for `CLight_ctor` the CLight class is now initialized directly
* Removed the `CLight_everyFrame` hook, lights are now drawn only when the `DRAW_LIGHT` native is invoked
* Added `PREPARE_LIGHT`, allowing users to easily initialize basic properties (similar to what `CREATE_LIGHT` did)

Example usage drawn directly:
```lua
CreateThread(function()
	while true do
		SetLightCoords(-50.32, 71.73, 72.79)
		DrawLight()
		Wait(0)
	end
end)
```

Example usage PREPARE_LIGHT:
```lua
local color = { r = 255, g = 255, b = 0 }
local intensity = 1.0
local radius = 2.0

CreateThread(function()
	while true do
		local camPos = GetFinalRenderedCamCoord()
		local endPoint, normal = GetWorldCoordFromScreenCoord(GetDisabledControlNormal(0, 239), GetDisabledControlNormal(0, 240))
		endPoint = endPoint + normal * 10.0

		local rayHandle = StartExpensiveSynchronousShapeTestLosProbe(camPos.x, camPos.y, camPos.z, endPoint.x, endPoint.y, endPoint.z, -1, PlayerPedId(), 0)
		local _, hit, worldPosition = GetShapeTestResultIncludingMaterial(rayHandle)

		if hit then
			PrepareLight(1, (1<<1) | (1<<12), worldPosition.x, worldPosition.y, worldPosition.z, color.r, color.g, color.b, intensity)
			SetLightRadius(radius)
			DrawLight()
		end
		Wait(0)
	end
end)
```

fishiidev commented:
As someone who's worked with the limited FiveM light environment before, I can confidently say this PR is a solid step forward.

Even in its revised form, the addition of `PREPARE_LIGHT` and the more flexible `DRAW_LIGHT` implementation opens up a ton of new use cases that just weren’t possible before—especially with support for features like volumetric lighting, capsule shapes, and texture projection.

While the handle system had potential for persistent lights and optimization, I understand the concerns about persistence after script unload and streaming limits. The updated approach still achieves the main goal: giving developers a modern, more powerful lighting toolkit without overcomplicating implementation or impacting performance unpredictably.

For quick flash effects, as mentioned in earlier comments, I agree that helper functions or a native handling flashiness could make this even more powerful—but that feels like a great next step rather than a blocker for this PR.

This is a meaningful quality-of-life improvement for the community, especially for those of us trying to push the engine's visuals. I strongly support merging this.

Lucas7yoshi commented:
Question/concern... Is this limited to one light in a given frame? That might greatly limit its usefullness (atleast for me)... I may be misunderstanding though
